testsuite/ui/navigation/step_processor.py

```diff
@@ -0,0 +1,84 @@
+
+
+def view_class(step, dest, **kwargs):
+    """
+    Method defining rule where
+    """
+    if inspect.isclass(step.destination) and step.destination == dest:
+        _invoke_step_method(step, dest, **kwargs)
+        return True
+    return False
+
+
+def view_class_string(step, dest, **kwargs):
+    if isinstance(step.destination, str) and step.destination == dest.__class__.__name__:
+        _invoke_step_method(step, dest, **kwargs)
+        return True
+    return False
+
+
+def href(step, dest, **kwargs):
+    if step.destination.startswith('@href'):
+        logging.debug(f"Navigator: invoking step method marked with `@href` {step}")
+        step.method(dest.path)
+        return True
+    return False
+
+
+def _invoke_step_method(step, destination_view, **kwargs):
+    """
+    Alternative steps begins with "@".
+    See `step(cls, **kwargs)` for alternatives.
+
+    :param alternative_steps: list of tuples [(method name, method)]. Describes all View methods
+        decorated with @step
+    :param destination: View class
+    :return: Bool value representing if step was preformed or not
+    """
+    try:
+        logging.debug(f"Navigator: invoking method {step} that leads to {destination_view}")
+        signature = inspect.signature(step.method)
+        filtered_kwargs = {key: value for key, value in kwargs.items() if key in signature.parameters}
+        bound = signature.bind(**filtered_kwargs)
+        bound.apply_defaults()
+        step.method(*bound.args, **bound.kwargs)
+    except Exception as exc:
+        raise NavigationStepException(step.method, destination_view) from exc
+
+
+class ViewStepsProcessor:
+    """
+    This class is responsible for processing step methods on current view with respect to
+    """
+
+    def __init__(self, current_view):
+        """
+
+        """
+        self.current_view = current_view
+        View_step = namedtuple('Step', 'method destination')
+        step_methods = [i[1] for i in inspect.getmembers(current_view, lambda o: hasattr(o, 'class_name'))]
+        self.view_steps = [View_step(method, method.class_name) for method in step_methods]
+
+    def invoke(self, desired_dest, **kwargs):
+        """
+        Perform one step. This method invokes method that is decorated with @step.
+        See `step(cls, **kwargs)` correct usage.
+
+        :param possible_steps: list of tuples [(method name, method)]. Describes all View methods
+            decorated with @step
+        :param destination: View class
+        :param kwargs: parameters that are passed to particular steps. E.g. account_id={id}
+        :return: Bool value representing if step was preformed or not
+        """
+        rules = [view_class, view_class_string, href]
+        logging.debug(f"Navigator: finding step from {self.current_view} to {desired_dest}")
+        for r in rules:
+            if [step for step in self.view_steps if r(step, desired_dest, **kwargs)]:
```

Comment:
I dont really understand what this if does. `any(r(step, desired_dest, **kwargs) for step in self.view_steps)`